cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,176 @@
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.services;
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.store.RunRecordDetail;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import org.apache.twill.api.RunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * System program management service for ensuring services are running/stopped as expected
+ */
+public class SystemProgramManagementService extends AbstractScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemProgramManagementService.class);
+
+  private final long scheduleInterval;
+  private final CConfiguration cConf;
+  private final ProgramRuntimeService programRuntimeService;
+  private final ProgramLifecycleService programLifecycleService;
+  private Map<ProgramId, Map<String, String>> enabledServices;
+
+  @Inject
+  SystemProgramManagementService(CConfiguration cConf, ProgramRuntimeService programRuntimeService,
+                                 ProgramLifecycleService programLifecycleService) {
+    this.cConf = cConf;
+    this.scheduleInterval = cConf.getLong(Constants.AppFabric.PROGRAM_SCAN_INTERVAL, 10L);
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  /**
+   * Sets the map of services that are currently enabled along with their runtime args.
+   * The services that are not present in map will be stopped during the next run of the service.
+   *
+   * @param enabledServices
+   */
+  public void setEnabledServices(Map<ProgramId, Map<String, String>> enabledServices) {
+    this.enabledServices = new HashMap<>(enabledServices);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (enabledServices == null) {
+      LOG.info("Services to run not yet set");
+      return;
+    }
+    try {
+      reconcileServices();
+    } catch (Exception exception) {
+      //catch here so a single iteration doesn't stop the service
+      LOG.error("Error in reconciling services", exception);
+    }
+  }
+
+  private void reconcileServices() {
+    //take a copy
+    Map<ProgramId, Map<String, String>> enabledServicesMap = new HashMap<>(this.enabledServices);
+    Set<ProgramId> enabledServices = enabledServicesMap.keySet();
+    List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = programRuntimeService.listAll(ProgramType.SERVICE);
```

Comment:
It is better to loop by the ProgramId and call `ProgramRuntimeService.list(ProgramId)` on each of them

Reply:
Listing with program id will return only the enabled services runs. We also need to identify the services to stop, so using listAll for service here.

Reply:
Since you are listing all runs, you can collect the list of extra runs to stop in here. Also, always collect the ProgramRunId to pin point to which runs to run.

Reply:
Done. Changes in https://github.com/cdapio/cdap/pull/12886